File: Part1_강의자료/Ch07_이분탐색/문제별코드/2110_공유기설치/2110.py
N, C = map(int,input().split())
li = sorted([int(input()) for _ in range(N)])

def calc_count(dist):
    past_x, cnt = li[0], 1
    for i in range(1, N):
        if li[i] - past_x >= dist:
            past_x = li[i]
            cnt += 1
    return cnt

l, r = 1, (li[N - 1] - li[0]) // (C - 1)
while l <= r:
    mid = (l + r) // 2
    if calc_count(mid) >= C:
        l = mid + 1
    else:
        r = mid - 1
print(r)

File: Part1_강의자료/Ch07_이분탐색/문제별코드/2110_공유기설치/2110.cpp
#include <algorithm>
#include <stdio.h>
#include <vector>
using namespace std;

int calc_count(vector<int> &vec, int dist)
{
    int past_x = vec[0];
    int cnt = 1;
    for (int i = 1; i < vec.size(); i++)
    {
        if (vec[i] - past_x >= dist)
        {
            past_x = vec[i];
            cnt++;
        }
    }
    return cnt;
}

int main()
{
    int N, C;
    scanf("%d %d", &N, &C);
    vector<int> vec(N);
    for (int i = 0; i < N; i++)
        scanf("%d", &vec[i]);
    sort(vec.begin(), vec.end());

    int l = 0, r = (vec[N - 1] - vec[0]) / (C - 1);
    while (l <= r)
    {
        int mid = (l + r) / 2;
        if (calc_count(vec, mid) >= C)
            l = mid + 1;
        else
            r = mid - 1;
    }
    printf("%d\n", r);
    return 0;
}
